velox/common/memory/MappedMemory.cpp

```diff
@@ -180,42 +218,43 @@ bool MappedMemoryImpl::allocate(
   throw std::runtime_error("Not implemented");
 }
 
-MachinePageCount MappedMemoryImpl::allocationSize(
+bool MappedMemoryImpl::allocateContiguous(
     MachinePageCount numPages,
-    MachinePageCount minSizeClass,
-    std::array<int32_t, kMaxSizeClasses>* sizeIndices,
-    std::array<int32_t, kMaxSizeClasses>* sizeCounts,
-    int32_t* numSizes) const {
-  int32_t needed = numPages;
-  int32_t pagesToAlloc = 0;
-  *numSizes = 0;
-  VELOX_CHECK_LE(
-      minSizeClass,
-      sizes_.back(),
-      "Requesting minimum size larger than largest size class");
-  for (int32_t sizeIndex = sizes_.size() - 1; sizeIndex >= 0; sizeIndex--) {
-    int32_t size = sizes_[sizeIndex];
-    bool isSmallest = sizeIndex == 0 || sizes_[sizeIndex - 1] < minSizeClass;
-    // If the size is less than 1/8 of the size from the next larger,
-    // use the next larger size.
-    if (size > (needed + (needed / 8)) && !isSmallest) {
-      continue;
-    }
-    int32_t numUnits = std::max(1, needed / size);
-    needed -= numUnits * size;
-    if (isSmallest && needed > 0) {
-      // If needed / size had a remainder, add one unit of smallest class.
-      numUnits++;
-      needed -= size;
+    Allocation* FOLLY_NULLABLE collateral,
+    ContiguousAllocation& allocation,
+    std::function<void(int64_t)> beforeAllocCB) {
+  MachinePageCount collateralSize = 0;
+  if (collateral) {
+    collateralSize = free(*collateral) / kPageSize;
+  }
+  auto largeCollateralSize = allocation.numPages();
+  if (largeCollateralSize) {
+    if (munmap(allocation.data(), allocation.size()) < 0) {
+      LOG(ERROR) << "munmap got " << errno << "for " << allocation.data()
+                 << ", " << allocation.size();
     }
-    (*sizeCounts)[*numSizes] = numUnits;
-    pagesToAlloc += numUnits * size;
-    (*sizeIndices)[(*numSizes)++] = sizeIndex;
-    if (needed <= 0) {
-      break;
+    allocation.reset(nullptr, nullptr, 0);
+  }
+  int64_t needed = numPages - collateralSize - largeCollateralSize;
+  if (beforeAllocCB) {
+    try {
+      beforeAllocCB(needed * kPageSize);
+    } catch (std::exception& e) {
+      beforeAllocCB(-(collateralSize + largeCollateralSize) * kPageSize);
+      numAllocated_ -= largeCollateralSize;
+      std::rethrow_exception(std::current_exception());
     }
   }
-  return pagesToAlloc;
+  numAllocated_.fetch_add(needed);
+  void* data = mmap(
+      nullptr,
+      numPages * kPageSize,
+      PROT_READ | PROT_WRITE,
+      MAP_PRIVATE | MAP_ANONYMOUS,
+      -1,
+      0);
+  allocation.reset(this, data, numPages * kPageSize);
```

Comment:
Is it important to re-use allocation parameter? Can this method simply return new ContiguousAllocation? This way the API would be easier to reason about.